pandas/core/indexes/base.py

```diff
@@ -3334,9 +3334,7 @@ def _union(self, other: Index, sort: bool | None):
 
         if (
             sort in (None, True)
-            and self.is_monotonic_increasing
-            and other.is_monotonic_increasing
-            and not (self.has_duplicates and other.has_duplicates)
+            and (self.is_unique or other.is_unique)
```

Comment:
I think the order here can have a difference. `is_monotonic_increasing` will provide `is_unique` "for free" if the index is both monotonic and unique and it hasn't already been cached.  

```
import pandas as pd
import numpy as np

arr = np.arange(1_000_000)

%timeit idx=pd.Index(arr); idx.is_unique and idx.is_monotonic_increasing
# 68.7 ms ± 5.13 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit idx=pd.Index(arr); idx.is_monotonic_increasing and idx.is_unique
# 2.91 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
Good catch.  i had expected the cache for all of these to get populated at the same time inside IndexEngine, will take a closer look

Reply:
OK, looks like engine.is_monotonic_increasing will populate the engine.is_unique cache but not vice-versa.  All of the paths within the engine that check is_unique do that check after checking self.is_monotonic_increasing.  im inclined to update the engine code so that the unique check always populates the is_monotonic_increasing cache in order to 1) make the perf not dependent on the order these are accessed and 2) avoid populating `engine.mapping` in these cases (which allocates memory)

Reply:
> im inclined to update the engine code so that the unique check always populates the is_monotonic_increasing cache in order to 1) make the perf not dependent on the order these are accessed and 2) avoid populating engine.mapping in these cases (which allocates memory)

that sounds good to me